ambuda/views/proofing/project.py

```diff
@@ -286,7 +326,68 @@ def search(slug):
     )
 
 
-@bp.route("/<slug>/replace")
+def _replace_text(project_, replace_form: ReplaceForm, query: str, replace: str):
+    """
+    Gather all matches for the "query" string and pair them the "replace" string.
+    """
+
+    results = []
+
+    query_pattern = re.compile(
+        query, re.UNICODE
+    )  # Compile the regex pattern with Unicode support
+
+    LOG.debug(f"Search/Replace text with {query} and {replace}")
+    for page_ in project_.pages:
+        if not page_.revisions:
+            continue
+        matches = []
+        latest = page_.revisions[-1]
+        LOG.debug(f"{__name__}: {page_.slug}")
+        for line_num, line in enumerate(latest.content.splitlines()):
+            if query_pattern.search(line):
+                try:
+                    marked_query = query_pattern.sub(
+                        lambda m: Markup(f"<mark>{escape(m.group(0))}</mark>"), line
+                    )
+                    marked_replace = query_pattern.sub(
+                        Markup(f"<mark>{escape(replace)}</mark>"), line
```

Comment:
> rather than build HTML directly, let's move all of this logic into the template. Then `marked_replace` can be a simple list of strings.

HTML and js take time for me to get right.  I'll try to find a way to do it.

Reply:
I understand that it's desirable to move HTML markup into Jinja, but it seems to complicate the solution somewhat.

When using query_patten.sub(), the matches are marked in-place with <mark> as they are found. If an unmarked line is passed to Jinja, we need to either send a list of matches in the line or re-run the regex in Jinja code. Additionally, to mark the replaced substrings accurately, they should be passed as a list. Therefore, marking these substrings in _replace_text() seems to strike a good balance between clean language separation and the simplicity of the solution.